tests/list.egg

```diff
@@ -0,0 +1,97 @@
+(datatype List
+	(Nil)
+	(Cons i64 List))
+
+(ruleset list)
+
+(function list-length (List) i64)
+(relation list-length-demand (List))
+(rule
+	((list-length-demand (Nil)))
+	((set (list-length (Nil)) 0))
+	:ruleset list)
+(rule
+	((list-length-demand (Cons head tail)))
+	((set (list-length-demand tail) ()))
+	:ruleset list)
+(rule
+	(	(list-length-demand (Cons head tail))
+		(= (list-length tail) tail-length))
+	((set (list-length (Cons head tail)) (+ tail-length 1)))
+	:ruleset list)
+
+(function list-get (List i64) i64)
+(relation list-get-demand (List i64))
+(rule
+	(	(list-get-demand list 0)
+		(= list (Cons head tail)))
+	((set (list-get list 0) head))
+	:ruleset list)
+(rule
+	(	(list-get-demand list n) (!= n 0)
+		(= list (Cons head tail)))
+	((list-get-demand tail (- n 1)))
+	:ruleset list)
+(rule
+	(	(list-get-demand list n)
+		(= list (Cons head tail))
+		(= item (list-get tail (- n 1))))
+	((set (list-get list n) item))
+	:ruleset list)
+
+(function list-append (List List) List)
+(relation list-append-demand (List List))
```

Comment:
For `list-append`, I actually think you don't need a demand relation. `list-append` itself can be used as demand, and then the semantics is just a few `rewrite`s like `(rewrite (list-append (Nil) list) list)`